File: pkg/dev_compiler/test/codegen_expected/language/abstract_method_test.js
dart_library.library('language/abstract_method_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__abstract_method_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const abstract_method_test = Object.create(null);
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  dart.copyProperties(abstract_method_test, {
    get length() {
      return dart.throw("error: top-level getter called");
    }
  });
  dart.copyProperties(abstract_method_test, {
    set height(x) {
      dart.throw("error: top-level setter called");
    }
  });
  abstract_method_test.width = function() {
    dart.throw("error: top-level function called");
  };
  dart.fn(abstract_method_test.width, VoidTodynamic());
  abstract_method_test.A = class A extends core.Object {
    get useLength() {
      return this.length;
    }
    setHeight(x) {
      return this.height = x;
    }
    useWidth() {
      return this.width();
    }
  };
  dart.setSignature(abstract_method_test.A, {
    methods: () => ({
      setHeight: dart.definiteFunctionType(dart.dynamic, [dart.dynamic]),
      useWidth: dart.definiteFunctionType(dart.dynamic, [])
    })
  });
  abstract_method_test.A1 = class A1 extends abstract_method_test.A {
    width() {
      return 345;
    }
    new(length) {
      this.length = length;
      this.height = null;
    }
  };
  dart.setSignature(abstract_method_test.A1, {
    constructors: () => ({new: dart.definiteFunctionType(abstract_method_test.A1, [core.int])}),
    methods: () => ({width: dart.definiteFunctionType(core.int, [])})
  });
  abstract_method_test.main = function() {
    let a = new abstract_method_test.A1(123);
    expect$.Expect.equals(123, a.useLength);
    a.setHeight(234);
    expect$.Expect.equals(234, a.height);
    expect$.Expect.equals(345, a.useWidth());
    core.print([a.useLength, a.height, a.useWidth()]);
  };
  dart.fn(abstract_method_test.main, VoidTodynamic());
  // Exports:
  exports.abstract_method_test = abstract_method_test;
});

File: tests/language_strong/abstract_method_test.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import "package:expect/expect.dart";

// Checks that abstract instance methods are correctly resolved.

int get length => throw "error: top-level getter called";
set height(x) {
  throw "error: top-level setter called";
}

width() {
  throw "error: top-level function called";
}

abstract class A {
  int get length; //    Abstract instance getter.
  set height(int x); // Abstract instance setter.
  int width(); //       Abstract instance method.

  // Must resolve to non-abstract length getter in subclass.
  get useLength => length;
  // Must resolve to non-abstract height setter in subclass.
  setHeight(x) => height = x;
  // Must resolve to non-abstract width() method in subclass.
  useWidth() => width();
}

class A1 extends A {
  int length; // Implies a length getter.
  int height; // Implies a height setter.
  int width() => 345;
  A1(this.length);
}

main() {
  var a = new A1(123);
  Expect.equals(123, a.useLength);
  a.setHeight(234);
  Expect.equals(234, a.height);
  Expect.equals(345, a.useWidth());
  print([a.useLength, a.height, a.useWidth()]);
}
